### PR TITLE
suppress matching ipv6 dns record

### LINKED ISSUE #3864: Cloudflare record content always causes updates even when the value doesn't change
### Confirmation

- [X] This is a bug with an existing resource and is not a feature request or enhancement. Feature requests should be submitted with Cloudflare Support or your account team.
- [X] I have searched the issue tracker and my issue isn't already found.
- [X] I have replicated my issue using the latest version of the provider and it is still present.

### Terraform and Cloudflare provider version

Terraform v1.8.5
on darwin_arm64
+ provider registry.terraform.io/cloudflare/cloudflare v4.40.0

### Affected resource(s)

- cloudflare_record

### Terraform configuration files

```hcl
locals {
  ipv6 = "my_ipv6"
}

resource "cloudflare_record" "my_record" {
  zone_id = cloudflare_zone.my_zone.id
  type    = "AAAA"
  name    = "v6"
  content = local.ipv6
  proxied = true
  ttl     = 1
}
```

### Link to debug output

https://gist.github.com/alexluong/aebb6f23c57805ff87316f8dd74667ac

### Panic output

_No response_

### Expected output

After the initial `terraform apply`, subsequent runs won't show any changes.

### Actual output

Terraform registers that the record is always changing, causing "cloudflare_record.my_record will be updated in-place" on every `terraform plan`.

### Steps to reproduce

1. Set up a new Cloudflare Record resource where the `content` is a variable (either a local variable or coming from another resource)
2. Run `terraform apply` to apply the change
3. Run `terraform apply` again immediately

### Additional factoids

_No response_

### References

Maybe this is fixed by #3776, I'm not sure. It seems this resource & field is actively being worked on.

### PR DESCRIPTION
**Pull request by alexluong**
Resolves #3864

Hi @jacobbednarz, here's the initial PR. I've gone over the contributing pages but there's a lot so please do let me know if there's anything else I should be adding. I'm also not sure if I should add a test here. I didn't see a unit test for record resource and wasn't sure how I can run acceptance safely. Would appreciate any guidance here.

I've built the provider and tested locally. It fixes my issue regarding IPv6.

Looking into the changelog thingy now.

### CODE REVIEW

**Comment by jacobbednarz** [comment]
thanks for the PR! can you ensure we add a regression test to cover this new functionality?

File: internal/sdkv2provider/resource_cloudflare_record.go [hunk 1]
```diff
@@ -501,3 +502,22 @@ func suppressTrailingDots(k, old, new string, d *schema.ResourceData) bool {
 
 	return strings.TrimSuffix(old, ".") == newTrimmed
 }
+
+func suppressMatchingIpv6(_, old, new string, _ *schema.ResourceData) bool {
```
Comment (jacobbednarz) [inline internal/sdkv2provider/resource_cloudflare_record.go:506]: if you're not using these extra parameters, just get rid of them.

```suggestion
func suppressMatchingIpv6(old, new string) bool {
```

**Comment by jacobbednarz** [comment]
that command looks right to me and i can run it locally. definitely make sure you don't have `CLOUDFLARE_API_TOKEN` or similar exported because that will mess with the auth scheme used.

as for proxyman to check, you'll probably need to enable SSL proxying on the domain to ensure that you see more than the CONNECT requests.

File: internal/sdkv2provider/resource_cloudflare_record_test.go [hunk 2]
```diff
@@ -672,6 +673,35 @@ func TestAccCloudflareRecord_ClearTags(t *testing.T) {
...
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: providerFactories,
+		CheckDestroy:      testAccCheckCloudflareRecordDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckCloudflareRecordConfigIPv6(zoneID, rnd, rnd, "2001:4860:4860:0:0:0:0:8888"),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(name, "name", rnd),
+					resource.TestCheckResourceAttr(name, "type", "AAAA"),
+					resource.TestCheckResourceAttr(name, "content", "2001:4860:4860:0:0:0:0:8888"),
```
Comment (jacobbednarz) [inline internal/sdkv2provider/resource_cloudflare_record_test.go:692]: the Cloudflare API returns this shortened as well so you'll want the assertion here to be the short version of `2001:4860:4860::8888`.

```suggestion
					resource.TestCheckResourceAttr(name, "content", "2001:4860:4860::8888"),
```

**Comment by alexluong** [comment]
Hmm I don't believe there's any exported env. I'll double check and confirm again later.

File: internal/sdkv2provider/resource_cloudflare_record_test.go (hunk 2 above)
Comment (jacobbednarz) [reply internal/sdkv2provider/resource_cloudflare_record_test.go:692]: with this assertion fix, the acceptance test suite is passing

```
TF_ACC=1 go test ./internal/sdkv2provider -v -run "^TestAccCloudflareRecord_" -count 1 -timeout 120m -parallel 1
=== RUN   TestAccCloudflareRecord_ImportBasic
--- PASS: TestAccCloudflareRecord_ImportBasic (3.05s)
=== RUN   TestAccCloudflareRecord_ImportSRV
--- PASS: TestAccCloudflareRecord_ImportSRV (2.88s)
=== RUN   TestAccCloudflareRecord_Basic
=== PAUSE TestAccCloudflareRecord_Basic
=== RUN   TestAccCloudflareRecord_BasicValue
=== PAUSE TestAccCloudflareRecord_BasicValue
=== RUN   TestAccCloudflareRecord_CaseInsensitive
=== PAUSE TestAccCloudflareRecord_CaseInsensitive
=== RUN   TestAccCloudflareRecord_Apex
=== PAUSE TestAccCloudflareRecord_Apex
=== RUN   TestAccCloudflareRecord_LOC
=== PAUSE TestAccCloudflareRecord_LOC
=== RUN   TestAccCloudflareRecord_SRV
=== PAUSE TestAccCloudflareRecord_SRV
=== RUN   TestAccCloudflareRecord_CAA
=== PAUSE TestAccCloudflareRecord_CAA
=== RUN   TestAccCloudflareRecord_Proxied
=== PAUSE TestAccCloudflareRecord_Proxied
=== RUN   TestAccCloudflareRecord_Updated
=== PAUSE TestAccCloudflareRecord_Updated
=== RUN   TestAccCloudflareRecord_typeForceNewRecord
=== PAUSE TestAccCloudflareRecord_typeForceNewRecord
=== RUN   TestAccCloudflareRecord_hostnameForceNewRecord
=== PAUSE TestAccCloudflareRecord_hostnameForceNewRecord
=== RUN   TestAccCloudflareRecord_CreateAfterManualDestroy
=== PAUSE TestAccCloudflareRecord_CreateAfterManualDestroy
=== RUN   TestAccCloudflareRecord_TtlValidation
=== PAUSE TestAccCloudflareRecord_TtlValidation
=== RUN   TestAccCloudflareRecord_ExplicitProxiedFalse
=== PAUSE TestAccCloudflareRecord_ExplicitProxiedFalse
=== RUN   TestAccCloudflareRecord_MXWithPriorityZero
=== PAUSE TestAccCloudflareRecord_MXWithPriorityZero
=== RUN   TestAccCloudflareRecord_TtlValidationUpdate
=== PAUSE TestAccCloudflareRecord_TtlValidationUpdate
=== RUN   TestAccCloudflareRecord_HTTPS
=== PAUSE TestAccCloudflareRecord_HTTPS
=== RUN   TestAccCloudflareRecord_SVCB
=== PAUSE TestAccCloudflareRecord_SVCB
=== RUN   TestAccCloudflareRecord_MXNull
=== PAUSE TestAccCloudflareRecord_MXNull
=== RUN   TestAccCloudflareRecord_DNSKEY
    provider_test.go:211: Skipping acceptance test for default zone (0da42c8d2132a9ddaf714f9e7c920711). Pending automating setup from https://developers.cloudflare.com/dns/dnssec/multi-signer-dnssec/.
--- SKIP: TestAccCloudflareRecord_DNSKEY (0.00s)
=== RUN   TestAccCloudflareRecord_ClearTags
=== PAUSE TestAccCloudflareRecord_ClearTags
=== RUN   TestAccCloudflareRecord_CompareIPv6
=== PAUSE TestAccCloudflareRecord_CompareIPv6
=== CONT  TestAccCloudflareRecord_Basic
--- PASS: TestAccCloudflareRecord_Basic (2.40s)
=== CONT  TestAccCloudflareRecord_CreateAfterManualDestroy
--- PASS: TestAccCloudflareRecord_CreateAfterManualDestroy (4.97s)
=== CONT  TestAccCloudflareRecord_CompareIPv6
--- PASS: TestAccCloudflareRecord_CompareIPv6 (3.09s)
=== CONT  TestAccCloudflareRecord_ClearTags
--- PASS: TestAccCloudflareRecord_ClearTags (4.02s)
=== CONT  TestAccCloudflareRecord_MXNull
--- PASS: TestAccCloudflareRecord_MXNull (2.08s)
=== CONT  TestAccCloudflareRecord_SVCB
--- PASS: TestAccCloudflareRecord_SVCB (2.32s)
=== CONT  TestAccCloudflareRecord_HTTPS
--- PASS: TestAccCloudflareRecord_HTTPS (2.04s)
=== CONT  TestAccCloudflareRecord_TtlValidationUpdate
--- PASS: TestAccCloudflareRecord_TtlValidationUpdate (2.83s)
=== CONT  TestAccCloudflareRecord_MXWithPriorityZero
--- PASS: TestAccCloudflareRecord_MXWithPriorityZero (2.30s)
=== CONT  TestAccCloudflareRecord_ExplicitProxiedFalse
--- PASS: TestAccCloudflareRecord_ExplicitProxiedFalse (5.58s)
=== CONT  TestAccCloudflareRecord_TtlValidation
--- PASS: TestAccCloudflareRecord_TtlValidation (0.29s)
=== CONT  TestAccCloudflareRecord_CAA
--- PASS: TestAccCloudflareRecord_CAA (4.64s)
=== CONT  TestAccCloudflareRecord_hostnameForceNewRecord
--- PASS: TestAccCloudflareRecord_hostnameForceNewRecord (4.52s)
=== CONT  TestAccCloudflareRecord_typeForceNewRecord
--- PASS: TestAccCloudflareRecord_typeForceNewRecord (4.94s)
=== CONT  TestAccCloudflareRecord_Updated
--- PASS: TestAccCloudflareRecord_Updated (4.52s)
=== CONT  TestAccCloudflareRecord_Proxied
--- PASS: TestAccCloudflareRecord_Proxied (2.64s)
=== CONT  TestAccCloudflareRecord_Apex
--- PASS: TestAccCloudflareRecord_Apex (2.55s)
=== CONT  TestAccCloudflareRecord_SRV
--- PASS: TestAccCloudflareRecord_SRV (2.75s)
=== CONT  TestAccCloudflareRecord_LOC
--- PASS: TestAccCloudflareRecord_LOC (2.53s)
=== CONT  TestAccCloudflareRecord_CaseInsensitive
--- PASS: TestAccCloudflareRecord_CaseInsensitive (2.93s)
=== CONT  TestAccCloudflareRecord_BasicValue
--- PASS: TestAccCloudflareRecord_BasicValue (3.83s)
PASS
ok  	github.com/cloudflare/terraform-provider-cloudflare/internal/sdkv2provider	74.214s
```

**Comment by jacobbednarz** [comment]
acceptance tests all passing

```
TF_ACC=1 go test ./internal/sdkv2provider -v -run "^TestAccCloudflareRecord_" -count 1 -timeout 120m -parallel 1
=== RUN   TestAccCloudflareRecord_ImportBasic
--- PASS: TestAccCloudflareRecord_ImportBasic (3.05s)
=== RUN   TestAccCloudflareRecord_ImportSRV
--- PASS: TestAccCloudflareRecord_ImportSRV (2.88s)
=== RUN   TestAccCloudflareRecord_Basic
=== PAUSE TestAccCloudflareRecord_Basic
=== RUN   TestAccCloudflareRecord_BasicValue
=== PAUSE TestAccCloudflareRecord_BasicValue
=== RUN   TestAccCloudflareRecord_CaseInsensitive
=== PAUSE TestAccCloudflareRecord_CaseInsensitive
=== RUN   TestAccCloudflareRecord_Apex
=== PAUSE TestAccCloudflareRecord_Apex
=== RUN   TestAccCloudflareRecord_LOC
=== PAUSE TestAccCloudflareRecord_LOC
=== RUN   TestAccCloudflareRecord_SRV
=== PAUSE TestAccCloudflareRecord_SRV
=== RUN   TestAccCloudflareRecord_CAA
=== PAUSE TestAccCloudflareRecord_CAA
=== RUN   TestAccCloudflareRecord_Proxied
=== PAUSE TestAccCloudflareRecord_Proxied
=== RUN   TestAccCloudflareRecord_Updated
=== PAUSE TestAccCloudflareRecord_Updated
=== RUN   TestAccCloudflareRecord_typeForceNewRecord
=== PAUSE TestAccCloudflareRecord_typeForceNewRecord
=== RUN   TestAccCloudflareRecord_hostnameForceNewRecord
=== PAUSE TestAccCloudflareRecord_hostnameForceNewRecord
=== RUN   TestAccCloudflareRecord_CreateAfterManualDestroy
=== PAUSE TestAccCloudflareRecord_CreateAfterManualDestroy
=== RUN   TestAccCloudflareRecord_TtlValidation
=== PAUSE TestAccCloudflareRecord_TtlValidation
=== RUN   TestAccCloudflareRecord_ExplicitProxiedFalse
=== PAUSE TestAccCloudflareRecord_ExplicitProxiedFalse
=== RUN   TestAccCloudflareRecord_MXWithPriorityZero
=== PAUSE TestAccCloudflareRecord_MXWithPriorityZero
=== RUN   TestAccCloudflareRecord_TtlValidationUpdate
=== PAUSE TestAccCloudflareRecord_TtlValidationUpdate
=== RUN   TestAccCloudflareRecord_HTTPS
=== PAUSE TestAccCloudflareRecord_HTTPS
=== RUN   TestAccCloudflareRecord_SVCB
=== PAUSE TestAccCloudflareRecord_SVCB
=== RUN   TestAccCloudflareRecord_MXNull
=== PAUSE TestAccCloudflareRecord_MXNull
=== RUN   TestAccCloudflareRecord_DNSKEY
    provider_test.go:211: Skipping acceptance test for default zone (0da42c8d2132a9ddaf714f9e7c920711). Pending automating setup from https://developers.cloudflare.com/dns/dnssec/multi-signer-dnssec/.
--- SKIP: TestAccCloudflareRecord_DNSKEY (0.00s)
=== RUN   TestAccCloudflareRecord_ClearTags
=== PAUSE TestAccCloudflareRecord_ClearTags
=== RUN   TestAccCloudflareRecord_CompareIPv6
=== PAUSE TestAccCloudflareRecord_CompareIPv6
=== CONT  TestAccCloudflareRecord_Basic
--- PASS: TestAccCloudflareRecord_Basic (2.40s)
=== CONT  TestAccCloudflareRecord_CreateAfterManualDestroy
--- PASS: TestAccCloudflareRecord_CreateAfterManualDestroy (4.97s)
=== CONT  TestAccCloudflareRecord_CompareIPv6
--- PASS: TestAccCloudflareRecord_CompareIPv6 (3.09s)
=== CONT  TestAccCloudflareRecord_ClearTags
--- PASS: TestAccCloudflareRecord_ClearTags (4.02s)
=== CONT  TestAccCloudflareRecord_MXNull
--- PASS: TestAccCloudflareRecord_MXNull (2.08s)
=== CONT  TestAccCloudflareRecord_SVCB
--- PASS: TestAccCloudflareRecord_SVCB (2.32s)
=== CONT  TestAccCloudflareRecord_HTTPS
--- PASS: TestAccCloudflareRecord_HTTPS (2.04s)
=== CONT  TestAccCloudflareRecord_TtlValidationUpdate
--- PASS: TestAccCloudflareRecord_TtlValidationUpdate (2.83s)
=== CONT  TestAccCloudflareRecord_MXWithPriorityZero
--- PASS: TestAccCloudflareRecord_MXWithPriorityZero (2.30s)
=== CONT  TestAccCloudflareRecord_ExplicitProxiedFalse
--- PASS: TestAccCloudflareRecord_ExplicitProxiedFalse (5.58s)
=== CONT  TestAccCloudflareRecord_TtlValidation
--- PASS: TestAccCloudflareRecord_TtlValidation (0.29s)
=== CONT  TestAccCloudflareRecord_CAA
--- PASS: TestAccCloudflareRecord_CAA (4.64s)
=== CONT  TestAccCloudflareRecord_hostnameForceNewRecord
--- PASS: TestAccCloudflareRecord_hostnameForceNewRecord (4.52s)
=== CONT  TestAccCloudflareRecord_typeForceNewRecord
--- PASS: TestAccCloudflareRecord_typeForceNewRecord (4.94s)
=== CONT  TestAccCloudflareRecord_Updated
--- PASS: TestAccCloudflareRecord_Updated (4.52s)
=== CONT  TestAccCloudflareRecord_Proxied
--- PASS: TestAccCloudflareRecord_Proxied (2.64s)
=== CONT  TestAccCloudflareRecord_Apex
--- PASS: TestAccCloudflareRecord_Apex (2.55s)
=== CONT  TestAccCloudflareRecord_SRV
--- PASS: TestAccCloudflareRecord_SRV (2.75s)
=== CONT  TestAccCloudflareRecord_LOC
--- PASS: TestAccCloudflareRecord_LOC (2.53s)
=== CONT  TestAccCloudflareRecord_CaseInsensitive
--- PASS: TestAccCloudflareRecord_CaseInsensitive (2.93s)
=== CONT  TestAccCloudflareRecord_BasicValue
--- PASS: TestAccCloudflareRecord_BasicValue (3.83s)
PASS
ok  	github.com/cloudflare/terraform-provider-cloudflare/internal/sdkv2provider	74.214s
```

**Comment by jacobbednarz** [comment]
cheers for getting this one over the line 🚀

**Comment by alexluong** [comment]
totally, and thanks for your support @jacobbednarz!

**Comment by alexluong** [comment]
As for the acceptance test, here's the 400 request:

```
POST /client/v4/zones/0da42c8d2132a9ddaf714f9e7c920711/dns_records HTTP/1.1
Host: api.cloudflare.com
User-Agent: terraform-provider-cloudflare/dev terraform-plugin-sdk terraform/1.8.5
Content-Length: 204
Content-Type: application/json
X-Auth-Email: terraform-acceptance-test@cfapi.net
X-Auth-Key: Secret
Accept-Encoding: gzip

{"created_on":"0001-01-01T00:00:00Z","modified_on":"0001-01-01T00:00:00Z","type":"AAAA","name":"jknduanbut","content":"2001:4860:4860:0:0:0:0:8888","zone_id":"0da42c8d2132a9ddaf714f9e7c920711","ttl":3600}
```

response

```
HTTP/1.1 400 Bad Request
Date: Tue, 03 Sep 2024 09:24:05 GMT
Content-Type: application/json
Transfer-Encoding: chunked
Connection: keep-alive
CF-Ray: 8bd4af4f48f38616-HKG
CF-Cache-Status: DYNAMIC
Set-Cookie: __cflb=0H28vgHxwvgAQtjUGUFqYFDiSDreGJnUzDRejqX1Pru; SameSite=Lax; path=/; expires=Tue, 03-Sep-24 11:54:06 GMT; HttpOnly
cf-auditlog-id: 0191b732-cd9f-75e8-a5c8-e29652753917
Vary: Accept-Encoding
Set-Cookie: __cf_bm=63tBfX7LqejyAbGn9Cw2mdp3HAltPaLiwYs6w1apSjw-1725355445-1.0.1.1-Ua33.HoeLRtvAIn0tfZFbQpTVxD7pHYecPuLTucHSr76KJOyf4XDsOnulWBeNKUmnU1Pdzm5YKZZTg1YoD_SOA; path=/; expires=Tue, 03-Sep-24 09:54:05 GMT; domain=.api.cloudflare.com; HttpOnly; Secure; SameSite=None
Set-Cookie: __cfruid=55dfe664b12b0caf6f68f19d982d615692c959f4-1725355445; path=/; domain=.api.cloudflare.com; HttpOnly; Secure; SameSite=None
Server: cloudflare

{"success":false,"errors":[{"code":10000,"message":"Authentication error"}]}
```

Do you see anything wrong with by chance?

Granted, the PR is already merged so we can move on too if you prefer to not spend too much time on it. We can certainly tackle it again in the off chance that I make another contribution later 😅

**Comment by jacobbednarz** [comment]
are you actually setting `X-Auth-Email` to the value above? that should be your own email if so.

**Comment by alexluong** [comment]
@jacobbednarz oops I was confused, I wasn't sure if I was supposed to use the exact value in the docs, sort of like a production mock, or if I should use my real credentials.

I did try my credentials too but it still didn't work. I assume my API key is invalid. I tested my TF setup, it uses `Authorization: Bearer <api_key>`. Can I use the same key for the `X-Auth-Key` header?

**Pull request merged**